runtime/lua/vim/lsp/rpc.lua

```diff
@@ -608,6 +619,83 @@ function M.connect(host, port)
   end
 end
 
+---@alias vim.lsp.rpc.Dispatcher fun(method: string, params: table<string, any>)
+---@alias vim.lsp.rpc.on_error fun(code: integer, ...: any)
+---@alias vim.lsp.rpc.on_exit fun(code: integer, signal: integer)
+
+---@class vim.lsp.rpc.PrivateDispatchers
+---@field notification vim.lsp.rpc.Dispatcher
+---@field server_request vim.lsp.rpc.Dispatcher
+---@field on_exit vim.lsp.rpc.on_error
+---@field on_error vim.lsp.rpc.on_exit
+
+-- TODO: better name
+---@class vim.lsp.rpc.StartServer
+---@field request fun(method: string, params: table?, callbackfun: fun(err: lsp.ResponseError | nil, result: any), notify_reply_callback:function?)
+---@field notify fun(method: string, params: any)
+---@field is_closing fun(): boolean
+---@field terminate fun(): nil
+
+--- Create a LSP RPC client factory that connects via named pipes (Windows)
+--- or unix domain sockets (Unix) to the given pipe_path (file path on
+--- Unix and name on Windows)
+---
+---@param pipe_path string file path of the domain socket (Unix) or name of the named pipe (Windows) to connect to
+---@return fun(dispatchers: vim.lsp.rpc.PrivateDispatchers): vim.lsp.rpc.StartServer? #function intended to be passed to |vim.lsp.rpc.start_client| or |vim.lsp.start| on the field cmd
+function M.domain_socket_connect(pipe_path)
+  return function(dispatchers)
+    dispatchers = merge_dispatchers(dispatchers)
+    local pipe, err_msg, err_name = uv.new_pipe(false)
+    if not pipe then
+      local _ = log.error()
+        and log.error(
+          'pipe with name %s could not be opened. Error: %s. Reason: %s',
+          pipe_path,
+          err_name,
+          err_msg
+        )
+      return
+    end
```

Comment:
I'd tend to wrap `uv.new_pipe(false)` into an `assert` - that will result in an error with trace for users.
There is no good way to recover from this and logging it would make it harder for users to figure out what went wrong.

Then the return type could also change to non-nullable.